domain/agentprovisioner/service/service.go

```diff
@@ -213,11 +214,46 @@ func (s *Service) ContainerConfig(ctx context.Context) (container.Config, error)
 	result.SnapStoreAssertions = modelConfig[config.SnapStoreAssertionsKey]
 	result.SnapStoreProxyID = modelConfig[config.SnapStoreProxyKey]
 	result.SnapStoreProxyURL = modelConfig[config.SnapStoreProxyURLKey]
-	_ = yaml.Unmarshal([]byte(modelConfig[config.CloudInitUserDataKey]), &result.CloudInitUserData)
 	result.ContainerInheritProperties = modelConfig[config.ContainerInheritPropertiesKey]
+
+	var ciud map[string]any
+	if err = yaml.Unmarshal([]byte(modelConfig[config.CloudInitUserDataKey]), &ciud); err != nil {
+		return container.Config{}, internalerrors.Capture(err)
+	}
+	result.CloudInitUserData = ensureStringKeyedMaps(ciud).(map[string]any)
+
 	return result, nil
 }
 
+// ensureStringKeyedMaps is used to ensure that maps in collections
+// with interface{} values are keyed by string if possible.
+// It processes the collection values recursively.
+func ensureStringKeyedMaps(data any) any {
```

Comment:
This seems like it would be a useful helper somewhere. Surely this isn't the only place we'd need to do this. Just a thought.